Core/Registry/Registry.cs

```diff
@@ -1027,45 +1027,65 @@ public List<string> GetSanityErrors()
         {
-            modules_to_remove = modules_to_remove.Memoize();
-            orig_installed    = orig_installed.Memoize();
+            modulesToRemove = modulesToRemove.Memoize();
+            origInstalled    = origInstalled.Memoize();
             var dllSet = dlls.ToHashSet();
             // The empty list has no reverse dependencies
             // (Don't remove broken modules if we're only installing)
-            if (modules_to_remove.Any())
+            if (modulesToRemove.Any())
             {
                 // All modules in the input are included in the output
-                foreach (string starter in modules_to_remove)
+                foreach (string starter in modulesToRemove)
                 {
                     yield return starter;
                 }
                 while (true)
                 {
                     // Make our hypothetical install, and remove the listed modules from it.
-                    HashSet<CkanModule> hypothetical = new HashSet<CkanModule>(orig_installed); // Clone because we alter hypothetical.
-                    hypothetical.RemoveWhere(mod => modules_to_remove.Contains(mod.identifier));
+                    HashSet<CkanModule> hypothetical = new HashSet<CkanModule>(origInstalled); // Clone because we alter hypothetical.
+                    if (modulesToInstall != null)
+                    {
+                        // Pretend the mods we are going to install are already installed, so that dependencies that will be
+                        // satisfied by a mod that is going to be installed count as satisfied.
+                        hypothetical = hypothetical.Concat(modulesToInstall).ToHashSet();
+                    }
+                    hypothetical.RemoveWhere(mod => modulesToRemove.Contains(mod.identifier));
 
-                    log.DebugFormat("Started with {0}, removing {1}, and keeping {2}; our dlls are {3}", string.Join(", ", orig_installed), string.Join(", ", modules_to_remove), string.Join(", ", hypothetical), string.Join(", ", dllSet));
+                    log.DebugFormat("Started with {0}, removing {1}, and keeping {2}; our dlls are {3}", string.Join(", ", origInstalled), string.Join(", ", modulesToRemove), string.Join(", ", hypothetical), string.Join(", ", dllSet));
 
                     // Find what would break with this configuration.
                     var broken = SanityChecker.FindUnsatisfiedDepends(hypothetical, dllSet, dlc)
                         .Select(x => x.Key.identifier).ToHashSet();
 
+                    if (modulesToInstall != null)
+                    {
+                        // We added modules to the hypothetical list which are not yet installed,
+                        // but not their dependencies (we don't know them yet).
+                        // Thus the SanityChecker wants to remove them again -> don't let him.
+                        broken = broken.Except(modulesToInstall.Select(m => m.identifier)).ToHashSet();
```

Comment:
Rather than removing the modules to install, could we take the intersection with `origInstalled` ? That would directly guarantee that `FindReverseDependencies` can only return modules that are already installed.

I keep thinking that there's still a way to make this freeze...

1. Install ModuleManager
2. In one change set:
   1. Install something that depends on something not yet installed that depends on ModuleManager
   2. Remove ModuleManager
3. Should say it's going to remove the intermediate dependency and freeze because it isn't installed yet

In actual testing this works fine, but I'm not sure why that is. I think we may still be leaving ourselves open to this in subtle ways and would like to see stronger protection against it.

Reply:
Hmm, I played this through, and it seems to work out:

Assuming A depends on B depends on MM.
`FindUnsatisfiedDepends()` only returns A (which is missing B), because B is not yet included (search for needed dependencies is done alter in the process).
A is removed from `broken` again because it is in `modulesToinstall` (yeah an intersection is cleaner), so broken is now empty.
Thus, `modulesToRemove` is a superset of `broken` and the loop stops.